docs/one-path-reachability-proofs.md

```diff
@@ -0,0 +1,174 @@
+=====================================================
+
+This document describes how to transform a given a one-step reachability goal
+by advancing it using a given rewrite axiom.
+
+This document *does not describe* how to select an axiom to be used for
+advancing the goal.
+
+
+Problem statement
+-----------------
+
+Consider the following "one-path" reachability goal
+
+```
+∀ X . φ(X) → ◆ ∃ Y . ψ(X, Y)
+```
+where `X` and `Y` are sets of variables and ◆ denotes "strong-eventually"
+
+Consider also the axiom
+
+`∀ Z . α(Z) → ⚪ β(Z)`.
+
+where ⚪  denotes "strong-next".
+
+The steps below show how the given goal is transformed by "applying" the 
+given axiom.
+
+Unrolling "strong-eventually"
+------------------------------
+
+Let `Δ = ν f . ⚪ f` denote the formula for diverging computation.
+Then,  for any formula ψ:
+
+- `◆ ψ = ◇ ψ ∨ Δ` by the definition of "strong eventually"
+- `    = ψ ∨ ⚪ ◇ ψ ∨ Δ` by unrolling `◆`
+- `    = ψ ∨ ⚪ ◇ ψ ∨ ⚪ Δ` by unrolling `Δ`
+- `    = ψ ∨ ⚪ (◇ ψ ∨ Δ)`
+  because `⚪` commutes with `∨` 
+- `    = ψ ∨ ⚪ ◆ ψ`
+
+
+Therefore proving
+```
+∀ X . φ(X) → ◆ ∃ Y . ψ(X, Y)
+```
+is equivalent to proving
+```
+∀ X . φ(X) →  ∃ Y . ψ(X, Y) ∨ ⚪ ◆ ∃ Y . ψ(X, Y)`
+```
+
+We can now move `∃ Y . ψ(X, Y)` to the right of the implication,
+and (assuming law of excluded middle) we obtain the equivalent:
+```
+∀ X . φ(X) ∧ ¬∃ Y . ψ(X, Y) →  ⚪ ◆ ∃ Y . ψ(X, Y)`
+```
+
+Applying the axiom `∀ Z . α(Z) → ⚪ β(Z)`
+------------------------------------------
+
+In the following let `Φ(X) = φ(X) ∧ ¬∃ Y . ψ(X, Y)`
+and `Ψ(X) = ◆ ∃ Y . ψ(X, Y)`
+ 
+
+We can now split the original goal onto `∃ Y. α(Y)` to obtain
+the equivalent goal:
+
+```
+(∀ X . Φ(X) ∧ (∃ Z. α(Z)) → ⚪ Ψ(X))
+∧
+(∀ X . Φ(X) ∧ (¬ ∃ Z. α(Z)) → ⚪ Ψ(X))
+```
+
+Now we know, from the basic symbolic execution algorithm that
+```
+Φ(X) ∧ (∃ Y .α(Y)) → ⚪ Φ'(X)    (Step)
+```
+for some `Φ'` determined by the algorithm.
+
+It is then sound to replace the first conjunct in the goal above with
+```
+∀ X . Φ'(X) → Ψ(X)
+```
+
+That is indeed so, because
+
+`∀ X . Φ'(X) → Ψ(X)` implies `∀ X . ⚪ Φ'(X) → ⚪ Ψ(X)`
+which, by using `(Step)` and implication transitivity, implies that
+```
+Φ(X) ∧ (∃ Y . α(Y)) → ⚪ Ψ(X)
+```
+
+
+Therefore, it is sound to replace the original goal with 
+```
+(∀ X . Φ'(X) → ◆ ∃ Y . ψ(X, Y))
+∧
+(∀ X . Φ(X) ∧ (¬ ∃ Z. α(Z)) → ⚪ ◆ ∃ Y . ψ(X, Y))
+```
+
+Examples of using the above procedure 
+-------------------------------------
+
+When applying an axiom on a symbolic configuration new constraints may be
+introduced on symbolic variables, coming from side conditions, or directly
+from the unification process.
+
+### Example (from side conditions)
+
+Consider proving that
+```
+forall A:Int . 3 / A -> \diamond \exists B:Int . B
+```
+
+using axiom
+```
+forall X:Int, Y:Int .  (Y =/=Int 0 = true) \and X / Y -> \next X /Int Y
+```
+
+Then the goal will simplify to
+```
+(\forall A:Int . 3 /Int A \and (A =/=Int 0 = true)-> \diamond \exists B:Int . B)
```

Comment:
I think it would be nice to have a form where Φ is visible, which then simplifies to the form that you have here, which uses φ.